gfx.cpp

```diff
@@ -4976,6 +5093,50 @@ class GfxInternal
         return gfx_texture.resource_state_;
     }
 
+    GfxResult setBufferResourceState(GfxBuffer const &buffer, D3D12_RESOURCE_STATES state)
+    {
+        if(!buffer_handles_.has_handle(buffer.handle))
+            return kGfxResult_InvalidParameter; // invalid buffer object
+        Buffer &gfx_buffer = buffers_[buffer];
+        if(transitionResource(gfx_buffer, state))
+            submitPipelineBarriers();
+        return kGfxResult_NoError;
+    }
+
+    GfxResult setTextureResourceState(GfxTexture const &texture, D3D12_RESOURCE_STATES state)
+    {
+        if(!texture_handles_.has_handle(texture.handle))
+            return kGfxResult_InvalidParameter; // invalid texture object
+        Texture &gfx_texture = textures_[texture];
+        if((state & D3D12_RESOURCE_STATE_UNORDERED_ACCESS) != 0)
+            ensureTextureHasUsageFlag(gfx_texture, D3D12_RESOURCE_FLAG_ALLOW_UNORDERED_ACCESS);
+        if((state & D3D12_RESOURCE_STATE_RENDER_TARGET) != 0)
+        {
+            uint32_t depth = texture.depth;
+            for(uint32_t i = 0; i < texture.mip_levels; ++i)
+            {
+                for(uint32_t j = 0; j < depth; ++j)
+                    ensureTextureHasRenderTargetView(texture, gfx_texture, i, j);
+                if(texture.is3D())
+                    depth = GFX_MAX(depth >> 1, 1u);
+            }
+        }
+        if((state & (D3D12_RESOURCE_STATE_DEPTH_WRITE | D3D12_RESOURCE_STATE_DEPTH_READ)) != 0)
+        {
+            uint32_t depth = texture.depth;
+            for(uint32_t i = 0; i < texture.mip_levels; ++i)
+            {
+                for(uint32_t j = 0; j < depth; ++j)
+                    ensureTextureHasDepthStencilView(texture, gfx_texture, i, j);
+                if(texture.is3D())
+                    depth = GFX_MAX(depth >> 1, 1u);
+            }
+        }
+        if(transitionResource(gfx_texture, state))
+            submitPipelineBarriers();
+        return kGfxResult_NoError;
+    }
```

Comment:
This uses the same code as used elsewhere to ensure that textures have the correct usage flag to allow the requested resource state to be transitioned to. `ensureTextureHasRenderTargetView` and `ensureTextureHasDepthStencilView` both call `ensureTextureHasUsageFlag` to ensure correct flags. These are the only flags that require handling within gfx.